// mx-rpc/src/runtime/RPCNetBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RPCNetBase = void 0;
const socket_io_1 = require("../net/socket.io");
const events_1 = require("events");
const proto_1 = require("../proto");
class RPCNetBase extends events_1.EventEmitter {
    constructor(role, req, srv, compress) {
        super();
        // inside: InsideClient = gInsideClient;
        this.clientSign = proto_1.gSystemSign;
        this.centerSign = "";
        this.compress = "";
        this.openCompress = false;
        this.onConnect = (function () {
            process.nextTick(this.send.bind(this), "REGIST", this.role, this.isReq, this.clientSign, this.compress);
        }).bind(this);
        this.isReq = req;
        this.role = role;
        if (typeof srv == "number") {
            this.net = new socket_io_1.RPCNetClient(srv);
        }
        else if (srv instanceof socket_io_1.RPCNetClient) {
            this.net = srv;
        }
        else {
            this.net = new socket_io_1.RPCNetClient();
        }
        this.compress = compress || "normal";
        this.net.on("data", this.onData.bind(this));
        this.net.on("disconnect", this._disconnect.bind(this));
        this.onConnect();
    }
    _disconnect() {
        // 这里标识网络断开了
        this.net.removeListener("connection", this.onConnect);
        this.net.once("connection", this.onConnect);
        console.log("relink");
    }
    _untilConnected(r, j) {
        if (this.net.connected || this.net.connectedF) {
            if (this.net.connected)
                r(true);
            else
                j();
        }
        else
            setTimeout(this._untilConnected.bind(this, r, j), 10);
    }
    init() {
        return new Promise((resolve, reject) => {
            this._untilConnected(resolve, reject);
        });
    }
    registret(sign, compress) {
        this.centerSign = sign;
        if (this.compress.length > 0 && this.compress == compress) {
            this.openCompress = true;
        }
    }
    send(...args) {
        // if (this.clientSign == this.centerSign) {
        //     return this.inside.send(...args);
        // }
        this.net.send(...args);
    }
    onData(type, requestID, route, ...args) {
        if (type == "REGISTRET") {
            this.registret(requestID, route);
            return;
        }
        // 这里route错了需要调整
        if (route == undefined || route == null || typeof route != "string") {
            console.error(...arguments);
            return;
        }
        let routes = route.split('.');
        // 确定一下角色正确性
        if (this.role != routes[0])
            return;
        if (this.listenerCount(type) != 0) {
            let content = {
                requestID: requestID,
                type: type,
                route: routes.slice(1, routes.length).join('.'),
                args: args,
            };
            this.emit(type, content);
        }
    }
}
exports.RPCNetBase = RPCNetBase;
