// mx-rpc/src/net/socket.io.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RPCNetServer = exports.RPCNetClient = void 0;
// 提供socket.io的支持，服务器之间使用socket.io 连接公共云层 云层按照链路找到实际的rpc服务
const socket_io_client_1 = __importDefault(require("socket.io-client"));
const socket_io_1 = __importDefault(require("socket.io"));
const events_1 = require("events");
var globalPort = 14001;
class RPCNetClient extends events_1.EventEmitter {
    constructor(port, host, proto) {
        super();
        this.connected = false;
        this.connectedF = false;
        let opt = {
            // forceNew: true,
            // autoConnect: true,
            transports: ['polling', 'websocket'],
            autoConnect: true
        };
        proto = proto || "http";
        host = host || "127.0.0.1";
        port = port || globalPort;
        if (port == 0 || port == 80 || port == 443) {
            this.conn = socket_io_client_1.default(`${proto}://${host}`, opt);
        }
        else {
            this.conn = socket_io_client_1.default(`${proto}://${host}:${port}`, opt);
        }
        this.conn.on("connect", this.onConnect.bind(this, true));
        this.conn.on("connect_error", this.onConnect.bind(this, false));
        this.conn.on("disconnect", this.onClose.bind(this));
        this.conn.on("message", this.onData.bind(this));
        this.conn.on("error", this.onData.bind(this));
        // 调整最大监听数量，多个业务服用一个套接字
        this.setMaxListeners(100);
        // this.conn.connect()
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
    onConnect(succ) {
        // console.log(arguments)
        succ ? this.connected = true : this.connectedF = true;
        this.emit("connection", succ);
    }
    onClose() {
        // console.log(arguments)
        this.emit("disconnect");
    }
    onData(...data) {
        // console.log(arguments)
        this.emit("data", ...data);
    }
    send(...args) {
        return this.conn.send(...args);
    }
}
exports.RPCNetClient = RPCNetClient;
class RPCNetServer extends events_1.EventEmitter {
    constructor(port) {
        super();
        let opt = {
            transports: ['websocket', "polling"],
            allowUpgrades: true
        };
        this.conn = socket_io_1.default.listen(port || globalPort, opt);
        this.conn.on("connection", this.onConnection.bind(this));
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
    onConnection(socket) {
        socket.on("message", this.onSocketData.bind(this, socket));
        socket.on("disconnect", this.onSocketClose.bind(this, socket));
        // console.log(arguments)
        this.emit("connection", socket.id);
    }
    onSocketData(socket, ...data) {
        // console.log(arguments)
        // 这里需要解析一下内容了 这里我们把传输过来的第一个参数作为 event 后续的作为参数
        let event = data[0];
        // console.log("onSocketData", JSON.stringify(data.slice(0, 3)))
        if (this.listenerCount(event) == 0) {
            this.emit("data", socket.id, ...data);
        }
        else {
            this.emit(event, socket.id, ...data.slice(1, data.length));
        }
    }
    onSocketClose(socket) {
        // console.log(arguments)
        this.emit("close", socket.id);
    }
    send(socketid, ...args) {
        let socket = this.conn.sockets.sockets[socketid];
        if (socket) {
            socket.send(...args);
        }
    }
}
exports.RPCNetServer = RPCNetServer;
// export function test() {
//     let s = new Server()
//     let c = new Client()
//     s.on("connection", function (ns) {
//         console.log(ns)
//     })
//     c.on("connection", function () {
//         console.log()
//     })
// }
